[sdba] Refactor interp_on_quantiles

aulemahal opened the pull request:
<!--Please ensure the PR fulfills the following requirements! -->
<!-- If this is your first PR, make sure to add your details to the AUTHORS.rst! -->
### Pull Request Checklist:
- [ ] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #xyz
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Documentation has been added / updated (for bug fixes / features)
- [x] HISTORY.rst has been updated (with summary of main changes)
  - [x] Link to issue (:issue:`number`) and pull request (:pull:`number`) has been added.
- [ ] `bumpversion patch` has been called on this branch
- [ ] The relevant author information has been added to `.zenodo.json`

### What kind of change does this PR introduce?

I though the changes in #914 would address a bug I saw while computing ESPO-R. However, I had forgotten about the splitting pathways for `group='time'` vs `group='time.prop'`. That also meant testing was not sufficient.  This PR changes `xclim.sdba.utils.interp_on_quantiles` so that:

* All NaNs in the coordinates are dropped. Minor modification from the Extremes' PR, but dropping everything explicitly, instead of relying on `interp1d`, has the advantage that NaNs in the new coord are not considered "out of bounds", but NaNs. They return NaN in the output, not an extrapolated value.
* All NaNs in the new coordinates are dropped. In the 2D case, with 'nearest', NaNs would cause QHull errors.

### Does this PR introduce a breaking change?
* Not really.

### Other information:
The docstring was improved.

I realized that `method='cubic'` does not respect the constant extrapolation as set by `extrapolate_qm`... I added a note to the doc, but I'm not sure what to do here. @huard ?  The logic we used works well with 'nearest' and 'linear', but as 'cubic' fetches a third point it estimates something else.

EDIT:
I compared the performance before and after the change and it is insignificant. `griddata` and `interp1d` are much slower than any numpy overhead I added, so I did not see any timing difference.

huard commented on xclim/sdba/utils.py:
```diff
@@ -417,35 +422,39 @@ def interp_on_quantiles(
+      Defaults to "time".
     method : {'nearest', 'linear', 'cubic'}
-      The interpolation method.
+      The interpolation method. Using cubic interpolation will NOT respect the constant
+      extrapolation as added by :py:func:`~xclim.sdba.utils.extrapolate_qm`.
```
You mean `interp_on_quantiles` will in fact extrapolate ? And this extrapolation is going to be cubic instead of constant. If the extrapolation used nan, then it's ok ? The cubic extrapolation will still be nan, right ?

aulemahal replied:
Yes, it will extrapolate. 'cubic' uses 3 points, It means the fitted spline for out-of-bounds data is not flat as it includes the second-to-last real point in addition to the last and its copy. I didn't try with `nan`, will do.

aulemahal replied:
Woups. When `extrapolate='nan', method='cubic'`, output is NaN everywhere.

Well. I'll fix that too.

aulemahal commented on xclim/sdba/utils.py:
```diff
@@ -459,31 +468,38 @@ def _interp_quantiles_1D(newx, oldx, oldy):
 
-        mask = np.isnan(_oldx) & np.isnan(_oldy) & np.isnan(_oldg)
-        mask[:, np.array([0, -1])] = False  # bounds might be NaN for a good reason.
+        if method != "nearest":
+            _oldx = np.clip(_oldx, _newx.min() - 1, _newx.max() + 1)
```
With non-nearest methods,  griddata uses `qhull` and constructs a "convex hull" with the coordinates before performing the interpolation. When we extrapolate with "constant", the bounds of `_oldx` are ±inf, and the convex hull creation fails.
I will add a comment.

aulemahal commented on xclim/sdba/utils.py:
```diff
@@ -459,31 +468,38 @@ def _interp_quantiles_1D(newx, oldx, oldy):
+        mask_old[:, np.array([0, -1])] = False  # bounds might be NaN for a good reason.
+        mask_old = mask_old | np.isnan(_oldx) | np.isnan(_oldg)
+
+        out = np.full_like(_newx, np.NaN, dtype=_oldy.dtype)
```
I'm not sure what interpolation on integer data would even mean, but I guess that is a possibility if the data is weird and the user unaware. I'll force a float dtype.

aulemahal commented:
Last commits went deeper and refactored `interp_on_quantiles`. 

- Extrapolation is now handled by that function, without need to use `extrapolate_qm`. Changes to the adjustment methods are internal and nothing has changed in the API for the normal user.
- However, extrapolation is now better handled. With `interp='cubic'`, the constant extrapolation is fixed.
- On a side note, I removed the `interp` argument of `Grouper`'s init. It served almost no purpose, introducing more noise than doing good. It stays on `Grouper.get_index` as boolean only. It only matters with monthly grouping, so... almost never. 
- A bit of code refactor.

Should we remove `extrapolate_qm`? It is not used in the adjustment methods anymore.

huard reviewed (approved):
Does it make sense to repurpose test_extrapolate_qm to test the new extrapolation code ? I know its implicitly tested elsewhere, but still.

aulemahal commented:
The new extrapolation code is explicitly tested in `test_interp_on_quantiles`!  (see `expe`).

I would rather suggest to _remove_ `extrapolate_qm`? I liked the way it dealt with the situation, but sadly `griddata` demands more complexity...

huard commented:
All good then. 
Ok for removal, I would be surprised anyone used that function directly. If we want to do it by the book, we could deprecate it for the next release before removal.
